compose: Eliminate the json/subscriptions/exists synchronous check.

alenavolk opened the pull request:
The discussion of these changes can be found [here](https://chat.zulip.org/#narrow/stream/frontend/subject/autosubscribe/near/456551).

To sum it up, a user is not allowed to send a message to the stream they are not subscribed to, unless the request contains the `autosubscribe` variable set to `true`. In this case, the user will be automatically subscribed to the stream before sending a message.

To see this behavior though the web interface, you can narrow down to a stream using the link of the following form: https://chat.zulip.org/?stream=frontend. The messages sent after narrowing down to a stream the regular way (by clicking on its name) will be rejected by the backend.

Please let me know if anything can be changed or improved.

Fixes #4650.

@showell @timabbott FYI :)

showell commented on static/js/compose.js:
```diff
@@ -561,28 +584,9 @@ exports.validate_stream_message_address_info = function (stream_name) {
-    return true;
+    var autosubscribe = page_params.narrow_stream !== undefined;
+    var error_type = check_unsubscribed_stream_for_send(stream_name, autosubscribe);
+    return validation_error(error_type, stream_name);
```
I think the function name here, `validation_error`, make you think the function doesn't have side effects, but it can cause `compose_error(..., $('stream')` as a side effect.

It might be cleaner to have the function just return a string, and then say something like this:

~~~ js
error_response = validation_error(error_type, stream_name)
if (error_response) {
    compose_error(error_response, $('#stream'));
}
return error_response

showell replied:
(I would also be fine with changing the name to something like `maybe_report_validation_error`.

alenavolk replied:
Addressed this in a separate commit. The function is renamed to `report_validation_error` because after we get rid of the synchronous check in ddf694e, it causes `compose_error` every time it's called.

showell commented on static/js/compose.js:
```diff
@@ -496,15 +496,8 @@ exports.get_invalid_recipient_emails = function () {
 
-function check_unsubscribed_stream_for_send(stream_name, autosubscribe) {
-    var stream_obj = stream_data.get_sub(stream_name);
+function check_unsubscribed_stream_for_send(stream_name) {
```
I'm a little confused about what happens to `autosubscribe` in this commit.

alenavolk replied:
In this commit (I think you're talking about the second one) I moved some of the code from `check_unsubscribed_stream_for_send` (the function we're planning to get rid of) to `validate_stream_message_address_info`, so passing `autosubscibe` as a parameter became unnecessary. Basically, this is a preliminary step to make further commits more readable.

showell commented:
@alenavolk Thanks for working on this!  I've started the review and added a few small comments.

One thing that's not completely clear to me is how we handle the case of somebody who has `page_params.narrow_stream` set but where the stream they're talking to is something that exists locally.  In that case, we'll want to use local echo, for example.

alenavolk commented:
@showell Thanks for reviewing this! I addressed some of your feedback, and planning to investigate what happens when the stream exists only locally a bit later.

timabbott commented:
@alenavolk it looks like this fails the node tests because it reduces coverage on `transmit.js`; can you investigate?

timabbott commented:
I merged the first commit, since I definitely like that refactor.

alenavolk commented:
@timabbott Sure, I'll let you know as soon as I fix that.

alenavolk commented:
@timabbott Sorry about the delay, the PR is ready for review.
